src/monitor/circuit_graph.rs

```diff
@@ -0,0 +1,379 @@
+    pub(super) fn is_strict_input(&self) -> bool {
+        matches!(self.kind, NodeKind::StrictInput { .. })
+    }
+
+    /// Returns `self.output` if `self` is a strict input operator.
+    pub(super) fn output_id(&self) -> Option<NodeId> {
+        if let NodeKind::StrictInput { output } = &self.kind {
+            Some(*output)
+        } else {
+            None
+        }
+    }
+
+    /// Generate unique name for the node to use as a node label in a visual
+    /// graph.
+    fn node_identifier(node_id: &GlobalNodeId) -> String {
+        let mut node_ident = "n".to_string();
+
+        for i in 0..node_id.path().len() {
+            node_ident.push_str(&node_id.path()[i].to_string());
+            if i < node_id.path().len() - 1 {
+                node_ident.push('_');
+            }
+        }
+        node_ident
+    }
+
+    /// Output circuit node as a node in a visual graph.
+    fn visualize<F>(&self, annotate: &F) -> Option<VisNode>
+    where
+        F: Fn(&GlobalNodeId) -> String,
+    {
+        match &self.kind {
+            NodeKind::Operator => {
+                let annotation = annotate(&self.id);
+                Some(VisNode::Simple(SimpleNode::new(
+                    Self::node_identifier(&self.id),
+                    format!(
+                        "{}{}{}",
+                        self.name,
+                        if annotation.is_empty() { "" } else { "\\l" },
+                        annotation
+                    ),
+                )))
+            }
+            NodeKind::Circuit { region, .. } => {
+                Some(VisNode::Cluster(region.visualize(self, annotate)))
+            }
+            NodeKind::StrictInput { .. } => None,
+        }
+    }
+}
+
+pub(super) struct CircuitGraph {
+    /// Tree of nodes.
+    nodes: Node,
+    /// Matches a node to the vector of nodes that read from its output
+    /// stream or have a dependency on it.
+    /// A node can occur in this vector multiple times.
+    edges: HashMap<GlobalNodeId, Vec<(GlobalNodeId, EdgeKind)>>,
+}
+
+impl CircuitGraph {
+    pub(super) fn new() -> Self {
+        Self {
+            nodes: Node::new(
+                GlobalNodeId::root(),
+                "root",
+                RegionId::root(),
+                NodeKind::Circuit {
+                    iterative: true,
+                    children: HashMap::new(),
+                    region: Region::new(RegionId::root(), Cow::Borrowed("root")),
+                },
+            ),
+            edges: HashMap::new(),
+        }
+    }
+
+    /// Locate node by its global id.
+    pub(super) fn node_ref(&self, id: &GlobalNodeId) -> Option<&Node> {
+        self.nodes.node_ref(id.path().iter())
+    }
+
+    /// Locate node by its global id.
+    pub(super) fn node_mut(&mut self, id: &GlobalNodeId) -> Option<&mut Node> {
+        self.nodes.node_mut(id.path().iter())
+    }
+
+    pub(super) fn add_edge(&mut self, from: &GlobalNodeId, to: &GlobalNodeId, kind: &EdgeKind) {
```

Comment:
This method should take ownership of these id vecs since it immediately clones them all